File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
int Sum(vector<int> &arr, int d){
	int sum=0;
	int n = arr.size();
	for(int i = 0;i<n;i++){
		sum = sum + ceil((double)arr[i]/ (double)d);

	}
	return sum;
}
    int smallestDivisor(vector<int>& arr, int limit) {
    int low = 1;
	int high = *max_element(arr.begin(), arr.end());
	while(low<=high){
		int mid = low + (high-low)/2;
		if(Sum(arr, mid)<=limit){
			high = mid-1;
		}else
			low =mid+1;
	}
	return low;
    }
};

File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.py
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        n = len(nums)

        maxi = max(nums)
        

        if n>threshold:
            return -1
        l = 1
        h = maxi
        res = maxi
        while l<=h:
            mid = (l+h)//2

            if divisor(nums,mid,threshold)<=threshold:
                res = min(res,mid)
                h = mid-1
            else:
                l = mid+1
        return res

def divisor(nums,div,threshold):
    n = len(nums)
    div_sum = 0
    for i in range(n):
        div_sum += math.ceil(nums[i]/div)
    return div_sum